server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/cache/CachedBlobStore.java

```diff
@@ -0,0 +1,165 @@
+        Preconditions.checkNotNull(bucketName, "bucketName should not be null");
+
+        return isDefaultBucketName(bucketName)
+            .flatMap(ignored ->
+                Mono.from(cache.read(blobId))
+                    .<InputStream>flatMap(bytes -> Mono.fromCallable(() -> new ByteArrayInputStream(bytes)))
+            )
+            .switchIfEmpty(Mono.fromCallable(() -> backend.read(bucketName, blobId)))
+            .blockOptional()
+            .orElseThrow(() -> new ObjectNotFoundException(String.format("Could not retrieve blob metadata for %s", blobId)));
+    }
+
+    @Override
+    public Mono<byte[]> readBytes(BucketName bucketName, BlobId blobId) {
+        return isDefaultBucketName(bucketName)
+            .flatMap(ignored -> Mono.from(cache.read(blobId)))
+            .switchIfEmpty(Mono.from(backend.readBytes(bucketName, blobId)));
+    }
+
+    @Override
+    public Mono<BlobId> save(BucketName bucketName, byte[] bytes, StoragePolicy storagePolicy) {
+        return Mono.from(backend.save(bucketName, bytes, storagePolicy))
+            .flatMap(blobId -> {
+                if (isAbleToCache(bucketName, bytes, storagePolicy)) {
+                    return Mono.from(cache.cache(blobId, bytes)).thenReturn(blobId);
+                }
+                return Mono.just(blobId);
+            });
+    }
+
+    @Override
+    public Publisher<BlobId> save(BucketName bucketName, InputStream inputStream, StoragePolicy storagePolicy) {
+        Preconditions.checkNotNull(inputStream, "InputStream must not be null");
+
+        PushbackInputStream pushbackInputStream = new PushbackInputStream(inputStream, sizeThresholdInBytes + 1);
+        return Mono.from(backend.save(bucketName, pushbackInputStream, storagePolicy))
+            .flatMap(blobId ->
+                Mono.fromCallable(() -> isALargeStream(pushbackInputStream))
+                    .flatMap(largeStream -> {
+                        if (!largeStream) {
+                            return Mono.from(saveInCache(bucketName, blobId, pushbackInputStream, storagePolicy))
+                                .thenReturn(blobId);
+                        }
+                        return Mono.just(blobId);
+                    })
+            );
+    }
+
+    @Override
+    public BucketName getDefaultBucketName() {
+        return defaultBucket;
+    }
+
+    @Override
+    public Mono<Void> delete(BucketName bucketName, BlobId blobId) {
+        return Mono.from(backend.delete(bucketName, blobId))
+            .then(isDefaultBucketName(bucketName)
+                .flatMap(ignored -> Mono.from(cache.remove(blobId)))
+                .then());
+    }
+
+    @Override
+    public Publisher<Void> deleteBucket(BucketName bucketName) {
+        return Mono.from(backend.deleteBucket(bucketName));
+    }
+
+    private Mono<BucketName> isDefaultBucketName(BucketName bucketName) {
```

Comment:
We sgould return boolean here, I see no reasons to wrap it in a mono.

(Callers can do `Mono.just` themselves)